### ISSUE TITLE
Error: wf.bold_surf_wf.rename_src

### ISSUE THREAD

**Issue by sundelinustc** [poldracklab/fmriprep#1261]
Dear fmriprep experts,

I met an error when preprocessing a sample data (1 T1 image, 6 runs of task images, and 1 run of resting state image) using fmriprep singularity version on a cluster: 

Errors
Node Name: fmriprep_wf.single_subject_01_wf.func_preproc_task_Predator_run_05_wf.bold_surf_wf.rename_src
Node Name: fmriprep_wf.single_subject_01_wf.func_preproc_task_Predator_run_03_wf.bold_surf_wf.rename_src
Node Name: fmriprep_wf.single_subject_01_wf.func_preproc_task_rest_wf.bold_surf_wf.rename_src
Node Name: fmriprep_wf.single_subject_01_wf.func_preproc_task_Predator_run_04_wf.bold_surf_wf.rename_src
Node Name: fmriprep_wf.single_subject_01_wf.func_preproc_task_Predator_run_01_wf.bold_surf_wf.rename_src
Node Name: fmriprep_wf.single_subject_01_wf.func_preproc_task_Predator_run_02_wf.bold_surf_wf.rename_src

May I know what problem it is and to resolve it?

Many thanks!

Bests,
Delin

**Comment by effigies** [comment]
Can you provide the version of fMRIPrep you're using and the output logs? It is unclear from what you have provided what error is being encountered.

**Comment by sundelinustc** [comment]
Thank you for the feedback!

We are running fMRIPrep 1.1.4 on a cluster using singularity.

The output log please see below. We tried different datasets but met the same errors on 90% of them.

Node: fmriprep_wf.single_subject_03_wf.func_preproc_task_rest_wf.bold_surf_wf.rename_src
Working directory: /mnt/Analyses/fMRIPrep/scratch-03/fmriprep_wf/single_subject_03_wf/func_preproc_task_rest_wf/bold_surf_wf/rename_src

Node inputs:

format_string = %(subject)s
in_file = <undefined>
keep_ext = True
parse_string = <undefined>
subject = ['fsaverage5']
use_fullpath = False

Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/pipeline/plugins/multiproc.py", line 308, in _send_procs_to_workers
    self.procs[jobid].run(updatehash=updatehash)
  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 471, in run
    result = self._run_interface(execute=True)
  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 1250, in _run_interface
    self.config['execution']['stop_on_first_crash'])))
  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 1172, in _collate_results
    (self.name, '\n'.join(msg)))
Exception: Subnodes of node: rename_src failed:
Subnode 0 failed
Error: Traceback (most recent call last):

  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/pipeline/engine/utils.py", line 106, in nodelist_runner
    result = node.run(updatehash=updatehash)

  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 471, in run
    result = self._run_interface(execute=True)

  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 555, in _run_interface
    return self._run_command(execute)

  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 635, in _run_command
    result = self._interface.run(cwd=outdir)

  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 523, in run
    outputs = self.aggregate_outputs(runtime)

  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 620, in aggregate_outputs
    raise error

  File "/usr/local/miniconda/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 613, in aggregate_outputs
    setattr(outputs, key, val)

  File "/usr/local/miniconda/lib/python3.6/site-packages/traits/trait_types.py", line 1493, in validate
    self.error( object, name, value )

  File "/usr/local/miniconda/lib/python3.6/site-packages/traits/trait_handlers.py", line 172, in error
    value )

traits.trait_errors.TraitError: The 'out_file' trait of a RenameOutputSpec instance must be a file name, but a value of '/home/ds366/linux/fsaverage5.nii' <class 'str'> was specified.

**Comment by effigies** [comment]
So this is a duplicate of #1016, which we thought had been resolved in 1.1.4. I'll have another look.

**Comment by effigies** [comment]
Can you share your full Singularity command line? Also, `echo $PYTHONPATH` from whatever environment you're submitting your jobs from.

**Comment by sundelinustc** [comment]
Thanks! The singularity command line please see below (we are submitting cluster jobs, but the fmriprep-related command line is below):
======================================
/usr/local/packages/singularity/bin/singularity run -B \
/mnt/BIAC/carr.dhe.duke.edu/Data/Lab/Delin/Predator:/mnt  \
/usr/local/packages/singularity/images/fmriprep.simg  \
/mnt/Origin /mnt/Analyses/fMRIPrep \
participant --participant-label sub-03 -w \
/mnt/Analyses/fMRIPrep/scratch-03 --omp-nthreads 6 -vv
======================================

I found it weird that the output of "echo $PYTHONPATH" is empty, therefore I tried another way
======================================
[ds366@blade17 Predator]$ python -c 'import sys; print(sys.path)'
['', '/usr/lib64/python27.zip', '/usr/lib64/python2.7', '/usr/lib64/python2.7/plat-linux2', '/usr/lib64/python2.7/lib-tk', '/usr/lib64/python2.7/lib-old', '/usr/lib64/python2.7/lib-dynload', '/usr/lib64/python2.7/site-packages', '/usr/lib64/python2.7/site-packages/gst-0.10', '/usr/lib64/python2.7/site-packages/gtk-2.0', '/usr/lib64/python2.7/site-packages/wx-2.8-gtk2-unicode', '/usr/lib/python2.7/site-packages', '/usr/lib/python2.7/site-packages/pymer4-0.0.1-py2.7.egg', '/usr/lib/python2.7/site-packages/line_profiler-2.1.2-py2.7-linux-x86_64.egg']
======================================

**Comment by sundelinustc** [comment]
[ds366@blade17 Predator]$ /usr/local/packages/singularity/bin/singularity run /usr/local/packages/singularity/images/fmriprep.simg --version
fmriprep v1.1.4
[ds366@blade17 Predator]$ /usr/local/packages/singularity/bin/singularity exec /usr/local/packages/singularity/images/fmriprep.simg python -c 'import nipype; print(nipype.__version__)'
1.1.2

**Comment by effigies** [comment]
Okay, thanks for all the information. I think this is going to have to be fixed in nipype. Let me open an issue there.

**Comment by sundelinustc** [comment]
Great thanks!

**Comment by effigies** [comment]
Sorry that took so long. I've given replicating it a couple tries. Do you think you could do the following:

```Shell
/usr/local/packages/singularity/bin/singularity exec /usr/local/packages/singularity/images/fmriprep.simg python
```

That should give you a prompt. You should then be able to copy and paste the example workflow I wrote in nipy/nipype#2687. If that fails in the same way, then we should be able to test solutions.

**Comment by sundelinustc** [comment]
Thank you so much! I tried to follow your instructions. Please see the below feedback:

[ds366@blade17 Predator]$ /usr/local/packages/singularity/bin/singularity exec /usr/local/packages/singularity/images/fmriprep.simg python
Python 3.6.5 |Anaconda, Inc.| (default, Apr 29 2018, 16:14:56) 
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> import numpy as np
>>> import nibabel as nb
>>> from nipype.pipeline import engine as pe
>>> from nipype.interfaces import utility as niu
>>> # needed values
... DEFAULT_MEMORY_MIN_GB = 0.01
>>> spaces = ['fsaverage5']
>>> nb.Nifti1Image(np.zeros((5,5,5,5), dtype=np.uint8),
...                np.eye(4)).to_filename('img.nii')
>>> rename_src = pe.MapNode(niu.Rename(format_string='%(subject)s', keep_ext=True),
...                         iterfield='subject', name='rename_src', run_without_submitting=True,
...                         mem_gb=DEFAULT_MEMORY_MIN_GB)
>>> rename_src.inputs.subject = spaces
>>> rename_src.inputs.in_file = os.path.abspath('img.nii')
>>> workflow = pe.Workflow(name='bold_surf_wf')
>>> workflow.add_nodes([rename_src])
>>> workflow.run(plugin='MultiProc')
180829-15:30:13,642 nipype.workflow INFO:
	 Workflow bold_surf_wf settings: ['check', 'execution', 'logging', 'monitoring']
180829-15:30:13,696 nipype.workflow INFO:
	 Running in parallel.
180829-15:30:13,710 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 1 jobs ready. Free memory (GB): 113.10/113.10, Free processors: 48/48.
180829-15:30:13,766 nipype.workflow INFO:
	 [Node] Setting-up "bold_surf_wf.rename_src" in "/tmp/tmp6t4gi4_h/bold_surf_wf/rename_src".
180829-15:30:13,769 nipype.workflow INFO:
	 [Node] Setting-up "_rename_src0" in "/tmp/tmp6t4gi4_h/bold_surf_wf/rename_src/mapflow/_rename_src0".
180829-15:30:13,771 nipype.workflow INFO:
	 [Node] Running "_rename_src0" ("nipype.interfaces.utility.base.Rename")
180829-15:30:13,777 nipype.workflow INFO:
	 [Node] Finished "_rename_src0".
180829-15:30:13,779 nipype.workflow INFO:
	 [Node] Finished "bold_surf_wf.rename_src".
180829-15:30:13,780 nipype.workflow INFO:
	 [Job 0] Completed (bold_surf_wf.rename_src).
<networkx.classes.digraph.DiGraph object at 0x2adae1393518>
>>>

**Comment by effigies** [comment]
Okay. That's what I get, too. So I'm very unclear on what's causing your fMRIPrep to consistently produce that specific failure. Could you post the full log? Perhaps there's another failure occurring.

**Comment by sundelinustc** [comment]
Please see the html report and the crash report attached:
[html.pdf](https://github.com/poldracklab/fmriprep/files/2332946/html.pdf)
[crash-20180824-211311-ds366-rename_src-08ff8723-3480-4c57-9472-c064c7086e81.txt](https://github.com/poldracklab/fmriprep/files/2332947/crash-20180824-211311-ds366-rename_src-08ff8723-3480-4c57-9472-c064c7086e81.txt)

**Comment by effigies** [comment]
Do you have the stdout for the run? The actual error may have occurred prior to this node, and it is only causing an exception at this point.

Also, does the following file exist on your filesystem: `/home/ds366/linux/fsaverage5.nii`

If so, can you run `ls -l /home/ds366/linux/fsaverage5.nii`?

**Comment by sundelinustc** [comment]
Thanks!

It seems that there are outputs for the functional images:
[ds366@blade17 Predator]$ ls Analyses/fMRIPrep/fmriprep/sub-03/func
sub-03_task-rest_bold_confounds.tsv
sub-03_task-rest_bold_space-MNI152NLin2009cAsym_brainmask.nii.gz
sub-03_task-rest_bold_space-MNI152NLin2009cAsym_preproc.nii.gz
sub-03_task-rest_bold_space-T1w_label-aparcaseg_roi.nii.gz
sub-03_task-rest_bold_space-T1w_label-aseg_roi.nii.gz

Moreover,
[ds366@blade17 Predator]$ ls -l /home/ds366/linux/fsaverage5.nii
ls: cannot access /home/ds366/linux/fsaverage5.nii: No such file or directory

**Comment by effigies** [comment]
Sorry for the confusion. [stdout](https://en.wikipedia.org/wiki/Standard_streams#Standard_output_(stdout)) refers to the output in the terminal. It may have been saved in a log file associated with the job by your batch system.

**Comment by sundelinustc** [comment]
Aha, sorry for the misunderstanding. We often submit cluster jobs using "qsub" function and get emails informing us the jobs have been completed (very simple, just the begin and end time). So, there is no terminal output. 
I will run the fmriprep commands in the terminal. But please wait for a few hours to get the final reports.
Many thanks!

**Comment by effigies** [comment]
`qsub` should cause files to be created of the form `<jobname>.['e'|'o']<job_id>'.'<task_id>`, where those with `o` are stdout, and `e` are stderr.

**Comment by aureliooo** [comment]
Here is the full log!

[Job Report.pdf](https://github.com/poldracklab/fmriprep/files/2333074/Job.Report.pdf)

**Comment by sundelinustc** [comment]
Yeah, my colleague aureliooo helped to find the job report as shown above.

**Comment by sundelinustc** [comment]
This problem seems to be resolved by my colleague aureliooo! It seems the working directory causes parallel processing of certain nodes and sometimes they interfere with each other and cause missing files. The solution is to set the working directory flag to "-w $TMPDIR". We tested it on at least 16 subjects and fmriprep ran smoothly with no errors.
